containerd: resource-limit settings for oci default

vyaghras opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #2814  #2862 

**Description of changes:**
Enable the capability to set the values for following resource limits
MaxAddressSpace,
MaxCoreFileSize,
MaxCpuTime,
MaxDataSize,
MaxFileLocks,
MaxFileSize,
MaxLockedMemory,
MaxMsgqueueSize,
MaxNicePriority,
MaxPendingSignals,
MaxProcesses,
MaxRealtimePriority,
MaxRealtimeTimeout,
MaxResidentSet,
MaxStackSize

**Testing done:**
- [x] Try setting a value for all the settings using api client -  It should show these values in etc/containerd/cri-base.json
- [x] Try setting a "unlimited" as hard-limit for max-locked-memory in the settings using api client -  It should show u64::MAX as hard limit for RLIMIT_MEMLOCK in etc/containerd/cri-base.json and the ulimit -Hl should set to  unlimited. 
- [x] Create a pod on bottlerocket host and check /proc/self/limits - It should list all the settings that has been done
         - Create a sample nginx workload container launched via kubernetes and check /proc/self/limits - This has default value for Max locked memory  as 65536.
         - Change Max locked memory setting using api client to 6536.
         - Create another container that runs the nginx image and check the /proc/self/limits - This has default value for Max locked memory  as 6536.
- [x] Migration testing 
         - Create an ami with 1.14.3 version and create instance using that.
         -  Try to set the max-locked-memory setting-  1.14.3 will not allow to do this setting.
         - Upgrade to 1.15.0 with the image in your custom TUF repo.
         - Try to set the max-locked-memory setting-  1.15.0 will allow to do memlock setting and add this in etc/containerd/cri_base.json.
         - Downgrade back to the older version 1.14.3.
         - Try to set the max-locked-memory setting-  1.14.3 will not allow to do this setting.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

vyaghras committed:
containerd: add all resource-limit settings for oci default

Add following rlimits settings for k8s variants, to enable updating
these from api client.
MaxAddressSpace,
MaxCoreFileSize,
MaxCpuTime,
MaxDataSize,
MaxFileLocks,
MaxFileSize,
MaxLockedMemory,
MaxMsgqueueSize,
MaxNicePriority,
MaxPendingSignals,
MaxProcesses,
MaxRealtimePriority,
MaxRealtimeTimeout,
MaxResidentSet,
MaxStackSize

stmcginnis reviewed (commented):
Looks like it's doing what's expected. I ran:

```sh
apiclient apply <<EOF
[settings.oci-defaults.resource-limits.max-address-space]
hard-limit = 2000000000
soft-limit = 1000000000

[settings.oci-defaults.resource-limits.max-core-file-size]
hard-limit = 2000000000
soft-limit = 1000000000

[settings.oci-defaults.resource-limits.max-cpu-time]
hard-limit = 10000
soft-limit = 1000

[settings.oci-defaults.resource-limits.max-data-size]
hard-limit = 1000000000
soft-limit = 100000000

[settings.oci-defaults.resource-limits.max-file-locks]
hard-limit = 2048
soft-limit = 1024

[settings.oci-defaults.resource-limits.max-file-size]
hard-limit = 204800
soft-limit = 102400

[settings.oci-defaults.resource-limits.max-locked-memory]
hard-limit = 1000000000
soft-limit = 100000000

[settings.oci-defaults.resource-limits.max-msgqueue-size]
hard-limit = 1000000001
soft-limit = 1000000000

[settings.oci-defaults.resource-limits.max-nice-priority]
hard-limit = 7
soft-limit = 6

[settings.oci-defaults.resource-limits.max-open-files]
hard-limit = 1048576
soft-limit = 65536

[settings.oci-defaults.resource-limits.max-pending-signals]
hard-limit = 1000
soft-limit = 500

[settings.oci-defaults.resource-limits.max-processes]
hard-limit = 10000
soft-limit = 9999

[settings.oci-defaults.resource-limits.max-realtime-priority]
hard-limit = 1000000000
soft-limit = 100000000

[settings.oci-defaults.resource-limits.max-realtime-timeout]
hard-limit = 1000000000
soft-limit = 100000000

[settings.oci-defaults.resource-limits.max-resident-set]
hard-limit = 1000000000
soft-limit = 100000000

[settings.oci-defaults.resource-limits.max-stack-size]
hard-limit = 1000000000
soft-limit = 100000000
EOF
```

Which results in `/etc/containerd/cri-base.json` containing:

```json
"rlimits": [{ "type": "RLIMIT_SIGPENDING", "hard": 1000, "soft": 500 },
{ "type": "RLIMIT_RTPRIO", "hard": 1000000000, "soft": 100000000 },
{ "type": "RLIMIT_RSS", "hard": 1000000000, "soft": 100000000 },
{ "type": "RLIMIT_NICE", "hard": 7, "soft": 6 },
{ "type": "RLIMIT_CORE", "hard": 2000000000, "soft": 1000000000 },
{ "type": "RLIMIT_NPROC", "hard": 10000, "soft": 9999 },
{ "type": "RLIMIT_AS", "hard": 2000000000, "soft": 1000000000 },
{ "type": "RLIMIT_RTTIME", "hard": 1000000000, "soft": 100000000 },
{ "type": "RLIMIT_NOFILE", "hard": 1048576, "soft": 65536 },
{ "type": "RLIMIT_STACK", "hard": 1000000000, "soft": 100000000 },
{ "type": "RLIMIT_CPU", "hard": 10000, "soft": 1000 }, 
{ "type": "RLIMIT_MEMLOCK", "hard": 1000000000, "soft": 100000000 },
{ "type": "RLIMIT_FSIZE", "hard": 204800, "soft": 102400 },
{ "type": "RLIMIT_MSGQUEUE", "hard": 1000000001, "soft": 1000000000 },
{ "type": "RLIMIT_LOCKS", "hard": 2048, "soft": 1024 },
{ "type": "RLIMIT_DATA", "hard": 1000000000, "soft": 100000000 }]
```

I do think you will need an `AddPrefixesMigration` migration though since these values didn't exist before. Might be good to confirm that with someone else that has a stronger grasp of our migration/datastore handling though.

cbgbt commented:
> I do think you will need an `AddPrefixesMigration`

That's correct. Consider the following workflow:
* You are running Bottlerocket X
* You upgrade to Y, which adds these new RLIMIT attributes to the model
* You decide to set `settings.oci-defaults.resource-limits.max-resident-set`
* You downgrade Bottlerocket from Y back to X. Version X of Bottlerocket doesn't know what `max-resident-set` is, so it fails to load the model.


An `AddPrefixesMigration` ensures that any new settings prefixes which won't be understood on older versions of Bottlerocket are removed on downgrade. You'd need to specify one of these migrations for each of the new types added (All in the same binary, `AddPrefixesMigration` accepts a list.)

vyaghras committed:
migrations: add migrations for 'settings.oci-defaults'

This migration will handle the added resource limits
settings added for k8s variants. The migration
oci-defaults-max-open-files handles the limit values
on upgrade and downgrade. If RLIMIT_NOFILE is unlimited
or more than u32::Max, then it will changed to max possible value in
version lesser than 1.15.0.

yeazelm commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -2737,4 +2737,14 @@ mod test_oci_spec {
     }
+
+    #[test]
+    fn oci_spec_mamlock_resource_limits_test() {
```
```suggestion
    fn oci_spec_memlock_resource_limits_test() {
```

yeazelm commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -125,5 +155,11 @@ mod oci_defaults_rlimits {
             "max-open-files",
             "RLIMIT_NOFILE",
         );
+
+        check_rlimit_strings(
+            OciDefaultsResourceLimitType::MaxLockedMemory,
+            "max-locked-memory",
+            "RLIMIT_MEMLOCK",
+        );
```
Is there a reason there is only this one new test and not one for all of the new settings we are adding? What is special about this setting?

vyaghras replied:
There is no strong reason for testing only this one but as this test is to check the mapping so I thought it will be redundant to test for all the settings.

yeazelm replied:
I find it a little odd to have only 2 unit tests for two of the sixteen settings we have here. Either testing that any one comes through correctly as a basic test makes sense or testing that each setting ends up where we want it makes sense, but just these two are hard to reason about. Its a lot of extra boilerplate but having one test for each setting is my preference since it would cover each case.

vyaghras replied:
Sure Its worth to have more checks always.. I will add one test for all 16 settings

etungsten commented on sources/api/migration/migrations/v1.15.0/oci-defaults-resource-setting/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added new resource limit settings for configuring the default OCI runtime spec,
+/// `settings.oci-defaults.resource-limits`, which will initially contain only
+/// `settings.oci-defaults.resource-limits.max-open-files`. We have to remove other
+/// settings on downgrade.
```
Should update this comment. Can just generically describe that we've added additional resource limits.

etungsten reviewed (approved):
nice! :rocket:

stmcginnis commented:
This can now be rebased on the latest `develop` to pick up a pin of the rust toolchain version used on the GitHub worker nodes. That should allow the checks to pass. This is a temporary workaround until the SDK is bumped to use rust 1.70.0.

yeazelm reviewed (approved):
This looks good. I do wonder if we do any validation on these values. For example, setting a soft limit above the hard limit is not possible or setting a value for nice outside of  -20 - 19. I'm not sure what we would do in those cases, but its worth considering if there is anything we would like to do there.

bcressey commented on Release.toml:
```diff
@@ -212,4 +212,6 @@ version = "1.15.0"
-"(1.14.1, 1.15.0)" = []
+"(1.14.1, 1.15.0)" = [
+    "migrate_v1.15.0_oci-defaults-resource-setting.lz4",
+]
```
after rebase, this now needs to be:
```suggestion
"(1.14.2, 1.15.0)" = [
    "migrate_v1.15.0_oci-defaults-resource-setting.lz4",
]
```

bcressey commented on sources/api/migration/migrations/v1.15.0/oci-defaults-resource-setting/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+/// We added new resource limit settings for configuring the default OCI runtime spec
+/// like max-address-space, max-core-file-size, max-cpu-time, max-data-size, max-file-locks
+/// max-file-size, max-locked-memory, max-msgqueue-size, max-nice-priority, max-pending-signals,
+/// max-processes, max-realtime-priority, max-realtime-timeout, max-resident-set, max-stack-size.
```
nit: no need to list all of these prefixes again:
```suggestion
/// We added new resource limit settings for configuring the default OCI runtime spec.
```

bcressey commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +150,100 @@ mod oci_defaults_rlimits {
+            OciDefaultsResourceLimitType::MaxAddressSpace,
+            "max-address-space",
+            "RLIMIT_AS",
+        );
```
nit: consider doing something like this:
```
        use OciDefaultsResourceLimitType::*;
        let rlimits = [
            (MaxAddressSpace, "max-address-space", "RLIMIT_AS"),
            (MaxCoreFileSize, "max-core-file-size", "RLIMIT_CORE"),
            ...
        ];

        for rlimit in rlimits {
            check_rlimit_strings(rlimit);
        }
```

With a slight tweak to the signature for `check_rlimit_strings` to take a tuple instead:
```
    fn check_rlimit_strings((cap, bottlerocket, linux): (OciDefaultsResourceLimitType, &str, &str)) { /* ... */ }
```

That way, the implementation matches the larger intent of calling the same function in the same way for each of the enum's variants. If some other function ever needs to be called, it will be obvious that it's an exception since it won't fit in the loop without special casing of some sort.

It also simplifies the reviewer / reader's job - it's just a question of whether the inputs are comprehensive, and not whether the right function is called in the right way at each call site.

bcressey commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -94,15 +94,45 @@ mod oci_defaults_capabilities {
 impl OciDefaultsResourceLimitType {
     pub fn to_linux_string(&self) -> &'static str {
         match self {
+            OciDefaultsResourceLimitType::MaxAddressSpace => "RLIMIT_AS",
```
nit: you can clean up some of the repetition with a `use` statement:
```suggestion
    pub fn to_linux_string(&self) -> &'static str {
        use OciDefaultsResourceLimitType::*;
        match self {
            MaxAddressSpace => "RLIMIT_AS",
```

bcressey commented on README.md:
```diff
@@ -749,10 +749,40 @@ Please see the [`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7
------ | ----- | -----
-`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.hard-limit` | 1048576
-`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.soft-limit` | 65536
+resource limit | setting | default value | unit
+----- | ----- | ----- | -----
+`RLIMIT_AS` | `settings.oci-defaults.resource-limits.max-address-space.hard-limit` | - | bytes
+`RLIMIT_AS` | `settings.oci-defaults.resource-limits.max-address-space.soft-limit` | - | bytes
```
If we use an HTML table instead, we could use `rowspan=2` for the resource limit and unit fields to avoid duplicating those values. Might be worth trying out to see if the result seems more polished.

bcressey commented on README.md:
```diff
@@ -749,10 +749,40 @@ Please see the [`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7
+`RLIMIT_NPROC` | `settings.oci-defaults.resource-limits.max-processes.soft-limit` | - | processes
+`RLIMIT_RTPRIO` | `settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit` | - | -
+`RLIMIT_RTPRIO` | `settings.oci-defaults.resource-limits.max-realtime-priority.soft-limit` | - | -
+`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit` | - | us
+`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit` | - | us
```
nit: `microseconds` might be more readable than `us`:
```suggestion
`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit` | - | microseconds
`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit` | - | microseconds
```

bcressey reviewed (commented):
It'd be nice to confirm that `-1` (for "infinity" / "unlimited") is accepted everywhere, or else to document what the right value to use is in order to get that behavior:

```
[settings.oci-defaults.resource-limits.max-address-space]
hard-limit = -1
soft-limit = -1
```

I'm not confident that `-1` will deserialize into `u32::MAX`, and I could see someone wanting to raise or remove a limit at runtime that they specified at launch time. That could be a problem since we don't allow settings to be deleted.

arnaldo2792 commented on README.md:
```diff
@@ -755,11 +755,190 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</tr>
+<tr>
+  <td rowspan="2">RLIMIT_AS</td>
+  <td>settings.oci-defaults.resource-limits.max-address-space.hard-limit</td>
```
You are missing " \`\` "  in all the settings. Same for `RLIMIT_<>`.

arnaldo2792 commented on README.md:
```diff
@@ -755,11 +755,190 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
-`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.soft-limit` | 65536
+<table>
+<tr>
+  <th>resource limit</th>
```
nit: capiltalize for each first letter

stmcginnis commented on README.md:
```diff
@@ -755,11 +755,190 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
------ | ----- | -----
-`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.hard-limit` | 1048576
-`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.soft-limit` | 65536
+<table>
```
It's highly unlikely we will want to render into anything other than HTML, but is there a reason for using HTML tags directly rather than sticking with markdown table syntax?

vyaghras replied:
So that we can use rowspan. This will avoid the repetition of `RLIMIT_NOFILE`  rlimit names.

bcressey commented on README.md:
```diff
@@ -755,11 +755,190 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+  <td rowspan="2"> <code>RLIMIT_AS</code> </td>
+  <td><code>settings.oci-defaults.resource-limits.max-address-space.hard-limit</code></td>
+  <td>-</td>
+  <td>bytes</td>
```
```suggestion
  <td rowspan="2">bytes</td>
```

bcressey reviewed (commented):
Code LGTM. One nit on the doc table, and I still have that open question about how to specify "unlimited" via settings, and making sure that works.

vyaghras commented:
> Code LGTM. One nit on the doc table, and I still have that open question about how to specify "unlimited" via settings, and making sure that works.

Now we can use "unlimited" string to set any rlimit as unlimited.

yeazelm commented on README.md:
```diff
@@ -755,12 +755,181 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</tr>
+</table>
+
+**Note**:  String `unlimited` can be used to set resource-limit as unlimited as
```
```suggestion
**Note**:  The string `unlimited` can be used to set resource-limit as unlimited. For example, to set the hard limit to unlimited:
```

yeazelm commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -2737,4 +2737,151 @@ mod test_oci_spec {
         );
     }
+
+    #[test]
+    fn oci_spec_max_address_space_resource_limits_test() {
+        let json = json!({"max-address-space": {"hard-limit": 3, "soft-limit": 4}});
+        let rendered = oci_spec_resource_limits(&json).unwrap();
+        assert_eq!(rendered, r#"{ "type": "RLIMIT_AS", "hard": 3, "soft": 4 }"#);
+    }
```
I'd like to see a test to confirm `unlimited` makes it through since we do some custom deserialization and e2e tests are a nice way to make sure things haven't shifted under us when making changes.

yeazelm commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +157,155 @@ mod oci_defaults_rlimits {
+    }
+}
+
+/// OciDefaultsResourceLimitValue specifies what all values are
+/// allowed to be set for rlimit in default spec.
```
We should call out `unlimited` and how this special case needs to have something more than a `u32`.

yeazelm commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +157,155 @@ mod oci_defaults_rlimits {
+                    )))
+                }
+            }
+            Value::String(s) => match s.to_lowercase().as_str() {
```
I have mixed feelings on doing `to_lowercase()` here. It's a balance between ease of use and correctness. I'm unfamiliar with if we typically do this elsewhere in settings but we should remain consistent with the rest of them.

yeazelm commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +157,155 @@ mod oci_defaults_rlimits {
+                    Ok(OciDefaultsResourceLimitValue::Value(s))
+                } else {
+                    Err(de::Error::custom(format!(
+                        "Provided integer is out of range: {}",
```
`Value::Number` can be a float too, so `1.0000` is in the range but not an `u64`. I would rather this error be more specific "Provided number is not within the range of 0 - {}: {}" and then put u64::MAX in there to provide the most correct error.

yeazelm commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +157,155 @@ mod oci_defaults_rlimits {
+            Value::Number(number) => {
+                if number.is_u64() {
+                    let s = number.to_string().parse::<u32>().map_err(|e| {
+                        de::Error::custom(format!(
```
I believe we use `Snafu` for these errors in `modeled_types` so I would like to use that for these errors.

vyaghras replied:
Serde expect errors while deserializing  in form of D::Error, hence we need to return the error in form of D::Error.

yeazelm replied:
These custom errors only in Serde deserializers makes sense.

bcressey commented on README.md:
```diff
@@ -755,12 +755,195 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+<tr>
+  <td><code>settings.oci-defaults.resource-limits.max-stack-size-limit</code></td>
+  <td>-</td>
+</tr>
```
```suggestion
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.soft-limit</code></td>
  <td>-</td>
</tr>
```

bcressey commented on README.md:
```diff
@@ -755,12 +755,195 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+<tr>
+  <td><code>settings.oci-defaults.resource-limits.max-open-files.soft-limit</code></td>
+  <td>65536</td>
+</tr>
```
nit: having "soft" come before "hard" would sort the numbers numerically, but should be done for all other limits too.

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -2737,4 +2737,160 @@ mod test_oci_spec {
+        );
+    }
+
+    #[test]
+    fn oci_spec_max_realtime_timeout_resource_limits_test() {
+        let json = json!({"max-realtime-timeout": {"hard-limit": 15, "soft-limit": 16}});
+        let rendered = oci_spec_resource_limits(&json).unwrap();
+        assert_eq!(
+            rendered,
+            r#"{ "type": "RLIMIT_RTTIME", "hard": 15, "soft": 16 }"#
+        );
+    }
```
nit: these could probably all be combined into one test that tests everything.

Typically when there are multiple test cases, I'd expect to see boundary testing - largest positive and smallest negative numbers, hard less than soft, that sort of thing - in addition to the test that exercises the happy path functionality.

I don't think you need more than two cases - one that checks that everything can be rendered, and one that checks for unlimited handling specifically.

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -516,8 +517,8 @@ struct OciDefaults {
 struct OciDefaultsResourceLimit {
-    hard_limit: u32,
-    soft_limit: u32,
+    hard_limit: OciDefaultsResourceLimitValue,
+    soft_limit: OciDefaultsResourceLimitValue,
```
This will require a data migration for the "max open files" limit - on downgrade, we need to turn `OciDefaultsResourceLimitValue` back into a `u32`, and on upgrade we might need to go the other way.

bcressey commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -120,10 +157,158 @@ mod oci_defaults_rlimits {
+    }
+}
+
+/// OciDefaultsResourceLimitValue specifies that any u32 integer and "unlimited"
+/// are the valid resource-limits. The hard-limit="unlimited" or
+/// soft-limit="unlimited" sets u64::MAX in spec file for container
+/// that converts to unlimited in containers for that resource
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum OciDefaultsResourceLimitValue {
+    Value(u32),
```
I am not confident about this - "max file size" seems to take a non-negative i64 value. You should check whether the same range is accepted for other limits.

We have a `NonNegativeInteger` type already (for i32) and could do the same for a `NonNegativeI64` type if needed.

If it turns out that `NonNegativeI64` is the right range for incoming values, then I would prefer to store everything as an i64, and represent "unlimited" as `-1`.

arnaldo2792 commented on README.md:
```diff
@@ -755,12 +755,195 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</tr>
+</table>
+
+Any integer between 0 to 4294967295 (2 ^32^ -1) and `unlimited` are valid values to set hard-limit and soft limit for a resource. 
```
nit: missing `-` in "soft limit"

arnaldo2792 commented on README.md:
````diff
@@ -755,12 +755,195 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+  [settings.oci-defaults.resource-limits.<rlimit>>]
+  hard-limit = 65536
+  soft-limit = 65536
+    ```
````
nit: There is an extra space here, and Github paints all of this red.

vyaghras commented:
Allow range of 0- i64::Max as limit and -1 and "unlimited" string to mark any limit as unlimited.

yeazelm commented on sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+/// There is no need of migration on upgrade as u32 will automatically change to i64
+pub struct ChangeMaxOpenFileResourceLimitType;
+
+fn convert_to_u_32(value: &mut Value) {
+    let s = if value.is_i64() {
+        let v: i64 = serde_json::from_value(value.clone()).unwrap();
+        match v {
+            -1 => u32::MAX,
+            v if v > u32::MAX as i64 => u32::MAX,
+            _ => v as u32,
+        }
+    } else {
+        return;
+    };
+    *value = Value::Number(s.into());
+}
```
This doesn't cover all cases. You need to either cover all cases that a `Value` can be, or limit the type to the ones you plan to cover (in this case `i64`). I think we are pretty certain this can only be a few things (mostly `i64` or `u32` since we are storing this in our datastore) so at least you could ensure its a `Number` but then you should also cover the case where could be a floating point. This is a little nitpicky but we should endeavor to use the type system when we can to avoid strange issues.

bcressey replied:
It should suffice to handle `value.is_i64()` and otherwise just leave the value alone, right?
```suggestion
fn convert_to_u_32(value: &mut Value) {
    if !value.is_i64() {
        return;
    }
    let v: i64 = serde_json::from_value(value.clone()).unwrap();
    match v {
        -1 => u32::MAX,
        v if v > u32::MAX as i64 => u32::MAX,
        _ => v as u32,
    }
    *value = Value::Number(s.into());
}
```

yeazelm replied:
I agree with @bcressey's suggestion.

yeazelm commented on sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+
+impl Migration for ChangeMaxOpenFileResourceLimitType {
+    /// On upgrade there is nothing to do (see above).
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        Ok(input)
```
If we did an end to end migration that was set as `unlimited`, would we end up with `unlimited` again or `u32::MAX`? This would be pretty surprising if its not getting us back to `unlimited`.

bcressey replied:
It feels like `u32::MAX` is the best we can do in that case - earlier versions didn't support anything beyond that, or "unlimited". Removing the limit entirely might cause it to be set much lower than expected, and `u32::MAX` open files is still quite a few.

yeazelm replied:
Yeah, I think I'm ok with `u32::MAX` over dropping it. I was considering if we should take `u32::MAX` on the forward case as `unlimited` but that might be less desirable than leaving it as we find it. Either way, I think the delta between `u32::MAX` and `unlimited` actually mattering in the real world is pretty low likelihood.

yeazelm commented on sources/models/src/de.rs:
```diff
@@ -207,3 +207,126 @@ mod mirrors_tests {
     }
 }
+
+// This specifies that any non negative i64 integer, -1 and "unlimited"
+/// are the valid resource-limits. The hard-limit="unlimited" or -1 and
+/// soft-limit="unlimited" or -1 sets u64::MAX in spec file for container
+/// that converts to unlimited in containers for that resource
```
```suggestion
// This specifies that any non negative i64 integer, -1, and "unlimited"
/// are the valid resource-limits. The hard-limit set to "unlimited" or -1 
/// and soft-limit set to "unlimited" or -1 are converted to u64::MAX in 
/// the spec file for the container runtime which ultimately represents 
/// unlimited for that resource
```

The last sentence was a little hard to understand, does this still mean the same thing or did I mis-understand what you are trying to say?

vyaghras replied:
Yeah it means same, will change to the one that you suggested.

bcressey commented on README.md:
```diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</tr>
+</table>
+
+Any integer between 0 to 9223372036854775807 (2 ^63^ -1), -1 and `unlimited` are valid values to set hard-limit and soft limit for a resource. 
```
```suggestion
Limits can be any integer between 0 to `int64::MAX`. Either `-1` or `"unlimited"` can be used to remove the limit.
```

bcressey commented on README.md:
```diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</table>
+
+Any integer between 0 to 9223372036854775807 (2 ^63^ -1), -1 and `unlimited` are valid values to set hard-limit and soft limit for a resource. 
+The string `unlimited` or -1 as resource-limit denotes limit as unlimited. Following are the examples to set hard-limit and soft-limit:
```
```suggestion
```

bcressey commented on README.md:
````diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+
+Any integer between 0 to 9223372036854775807 (2 ^63^ -1), -1 and `unlimited` are valid values to set hard-limit and soft limit for a resource. 
+The string `unlimited` or -1 as resource-limit denotes limit as unlimited. Following are the examples to set hard-limit and soft-limit:
+* Setting any valid integer value as limit:
+  ```toml
+  [settings.oci-defaults.resource-limits.<rlimit>>]
+  hard-limit = 65536
+  soft-limit = 65536
+  ```
````
This is implied by the next example.
```suggestion
```

bcressey commented on README.md:
````diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+  hard-limit = 65536
+  soft-limit = 65536
+  ```
+* Setting max possible integer value as limit:
+  ```toml
+  [settings.oci-defaults.resource-limits.<rlimit>>]
+  hard-limit = 9223372036854775807
+  soft-limit = 65536
+  ```
````
```suggestion
* Specifying the maximum value (`u64::MAX`) for a limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>]
  soft-limit = 65536
  hard-limit = 9223372036854775807
  ```
```

bcressey commented on README.md:
````diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+  hard-limit = 9223372036854775807
+  soft-limit = 65536
+  ```
+* Setting unlimited value as limit:
+  ```toml
+  [settings.oci-defaults.resource-limits.<rlimit>>]
+  hard-limit = "unlimited"
+  soft-limit = 65536
````
```suggestion
* Removing a limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>]
  soft-limit = 65536
  hard-limit = "unlimited"
```

bcressey commented on README.md:
````diff
@@ -756,12 +756,201 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+  hard-limit = "unlimited"
+  soft-limit = 65536
+  ```
+* Setting unlimited value as limit:
+  ```toml
+  [settings.oci-defaults.resource-limits.<rlimit>>]
+  hard-limit = -1
+  soft-limit = 65536
+  ```
````
We can support this in code, but don't need to document it. The "unlimited" one is more obvious and should be preferred.

bcressey commented on sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+/// There is no need of migration on upgrade as u32 will automatically change to i64
+pub struct ChangeMaxOpenFileResourceLimitType;
+
+fn convert_to_u_32(value: &mut Value) {
```
nit:
```suggestion
fn convert_to_u32(value: &mut Value) {
```

bcressey commented on sources/models/src/de.rs:
```diff
@@ -207,3 +207,126 @@ mod mirrors_tests {
+                Ok(-1)
+            } else {
+                Err(Error::custom(format!(
+                    "Invalid rlimit {}, use integer within the range of -1 - {} or string \" unlimited \" to set rlimit unlimited",
```
```suggestion
                    "Invalid rlimit {}, expected -1 to {} or \"unlimited\"",
```

bcressey commented on sources/models/src/de.rs:
```diff
@@ -207,3 +207,126 @@ mod mirrors_tests {
+                Ok(i)
+            } else {
+                Err(Error::custom(format!(
+                    "Provided number {} is not within the range of -1 - {}",
```
This can use the same error as above.

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -530,9 +530,28 @@ struct OciDefaults {
+}
+
+impl OciDefaultsResourceLimit {
+    pub fn get_hard_limit(self) -> u64 {
+        match self.hard_limit {
+            -1 => u64::MAX,
+            _ => self.hard_limit as u64,
+        }
+    }
```
nit: the body could be factored into a private `get_limit` function that takes either `self.hard_limit` or `self.soft_limit` as a parameter

bcressey reviewed (commented):
Very nice. Some of the documentation and code can be tightened up to convey the essential information and avoid repetition, but the logic looks solid.

bcressey commented on README.md:
```diff
@@ -756,12 +756,189 @@ Each of the `resource-limits` settings below contain two numeric fields: `hard-l
+</table>
+
+Limits can be any integer between 0 to `int64::MAX`. Either `-1` or `"unlimited"` can be used to remove the limit.
+* Specifying the maximum value (`u64::MAX`) for a limit:
```
```suggestion
* Specifying the maximum value (`i64::MAX`) for a limit:
```

bcressey commented on sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+    /// On downgrade, if the value is an i64 integer, we need to convert it to a u32.
+    ///
+    /// Note that this potentially causes data loss, if current value of the setting
+    /// is -1 or higher than u_32::MAX we  will set it to max possible value i.e. u32::MAX.
```
```suggestion
    /// is -1 or higher than u_32::MAX we will set it to max possible value i.e. u32::MAX.
```

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -530,9 +530,29 @@ struct OciDefaults {
+            -1 => u64::MAX,
+            _ => limit as u64,
+        }
+    }
```
nit: I forget where I picked up this convention, maybe Java, but I prefer to put public functions first since they're the external interface for the type, and private functions later since they're implementation details.

bcressey reviewed (approved):
LGTM with a few nits.

vyaghras commented:
I have tested this change again with all the migrations. The change is working as expected.
